core/src/main/scala/chisel3/internal/firrtl/IR.scala

```diff
@@ -743,8 +743,13 @@ abstract class Definition extends Command {
 case class DefPrim[T <: Data](sourceInfo: SourceInfo, id: T, op: PrimOp, args: Arg*) extends Definition
 case class DefInvalid(sourceInfo: SourceInfo, arg: Arg) extends Command
 case class DefWire(sourceInfo: SourceInfo, id: Data) extends Definition
-case class DefReg(sourceInfo: SourceInfo, id: Data, clock: Arg) extends Definition
-case class DefRegInit(sourceInfo: SourceInfo, id: Data, clock: Arg, reset: Arg, init: Arg) extends Definition
+case class DefReg(sourceInfo: SourceInfo, id: Data, clock: Arg, reset: Option[Arg], init: Option[Arg]) extends Definition {
+  assert(
+    (reset.isDefined && init.isDefined) || (reset.isEmpty && init.isEmpty),
+    "reset and init should exist(a register with init) or non-exist(a register without init) at the same time."
+  )
+  val hasInit: Boolean = reset.isDefined
+}
```

Comment:
The alternative formulation that avoids the need for a check is: `reset: Option[(Arg, Arg)]`.

Reply:
If you want names for the `reset` and `init` fields of the tuple, you could use a case case like:
```scala
case class RegInit(reset: Arg, init: Arg)
```
Then us an `Option[RegInit]`.

Reply:
`RegInit` is conflict with `chisel3.RegInit`, I use `RegInitIR` here.